src/cmfd_solver.cpp

```diff
@@ -34,8 +40,203 @@ xt::xtensor<int, 2> indexmap;
+  MPI_Reduce(&outside_, outside, 1, MPI_C_BOOL, MPI_LOR, 0, mpi::intracomm);
+
+#else
+  std::copy(cnt.data(), cnt.data() + total, cnt_reduced);
+  *outside = outside_;
+#endif
+
+  // Adapt reduced values in array back into an xarray
+  auto arr = xt::adapt(cnt_reduced, total, xt::acquire_ownership(), cnt_shape);
+  xt::xarray<double> counts = arr;
+
+  return counts;
+}
+
+//==============================================================================
+// OPENMC_CMFD_REWEIGHT performs reweighting of particles in source bank
+//==============================================================================
+
+extern "C"
+void openmc_cmfd_reweight(const bool feedback, const double* cmfd_src)
+{
+  // Get size of source bank and cmfd_src
+  auto bank_size = simulation::source_bank.size();
+  std::size_t src_size = cmfd::nx * cmfd::ny * cmfd::nz * cmfd::ng;
+
+  // count bank sites for CMFD mesh, store bins in bank_bins for reweighting
+  xt::xtensor<int, 1> bank_bins({bank_size}, 0);
+  bool sites_outside;
+  xt::xtensor<double, 1> sourcecounts = count_bank_sites(bank_bins,
+                                                         &sites_outside);
+
+  // Compute CMFD weightfactors
+  xt::xtensor<double, 1> weightfactors = xt::xtensor<double, 1>({src_size}, 1.);
+  if (mpi::master) {
+    if (sites_outside) {
+      fatal_error("Source sites outside of the CMFD mesh");
+    }
+
+    double norm = xt::sum(sourcecounts)()/cmfd::norm;
+    for (int i = 0; i < src_size; i++) {
+      if (sourcecounts[i] > 0 && cmfd_src[i] > 0) {
+        weightfactors[i] = cmfd_src[i] * norm / sourcecounts[i];
+      }
+    }
+  }
+
+  if (!feedback) {
+    return;
+  }
+
+#ifdef OPENMC_MPI
+  // Send weightfactors to all processors
+  MPI_Bcast(weightfactors.data(), src_size, MPI_DOUBLE, 0, mpi::intracomm);
+#endif
+
+  // Iterate through fission bank and update particle weights
+  for (int64_t i = 0; i < bank_size; i++) {
+    auto& site = simulation::source_bank[i];
+    site.wgt *= weightfactors(bank_bins(i));
+  }
+}
+
+//==============================================================================
+// OPENMC_INITIALIZE_MESH_EGRID sets the mesh and energy grid for CMFD reweight
+//==============================================================================
+
+extern "C"
+void openmc_initialize_mesh_egrid(const int meshtally_id, const int* cmfd_indices,
+                                  const double norm)
+{
+  // Make sure all CMFD memory is freed
+  free_memory_cmfd();
+
+  // Set CMFD indices
+  cmfd::nx = cmfd_indices[0];
+  cmfd::ny = cmfd_indices[1];
+  cmfd::nz = cmfd_indices[2];
+  cmfd::ng = cmfd_indices[3];
+
+  // Set CMFD reweight properties
+  cmfd::norm = norm;
+
+  // Find index corresponding to tally id
+  int32_t tally_index;
+  openmc_get_tally_index(meshtally_id, &tally_index);
+
+  // Get filters assocaited with tally
+  auto tally_filters = model::tallies[tally_index]->filters();
```

Comment:
Using `auto` will result in a copy. To avoid this, make it a reference, and since you are not changing it make it `const`:
```suggestion
  const auto& tally_filters = model::tallies[tally_index]->filters();
```